src/coreComponents/functions/FunctionBase.hpp

```diff
@@ -137,76 +142,77 @@ class FunctionBase : public dataRepository::Group
   localIndex totalVarSize = 0;
-  for( auto varIndex=0; varIndex<numVars; ++varIndex )
+  for( integer varIndex = 0; varIndex < numVars; ++varIndex )
   {
-    string const & varName = inputVarNames[varIndex];
+    string const & varName = m_inputVarNames[varIndex];
 
-    if( varName=="time" )
+    if( varName == "time" )
     {
-      input_ptrs[varIndex] = &time;
+      inputPtrs[varIndex] = &time;
       varSize[varIndex] = 1;
-      timeVar[varIndex] = 0;
-      ++totalVarSize;
     }
-    else if( groupSize > 0 )
+    else
     {
-      // Should we throw a warning if the group is zero-length?
       dataRepository::WrapperBase const & wrapper = group.getWrapperBase( varName );
-      input_ptrs[ varIndex ] = reinterpret_cast< double const * >( wrapper.voidPointer() );
+      varSize[varIndex] = wrapper.numArrayComp();
 
-      localIndex wrapperSize = LvArray::integerConversion< localIndex >( wrapper.size() );
-      varSize[varIndex] = wrapperSize / groupSize;
-      totalVarSize += varSize[varIndex];
+      using Types = types::ArrayTypes< types::RealTypes, types::DimsUpTo< 2 > >;
+      types::dispatch( Types{}, wrapper.getTypeId(), true, [&]( auto array )
+      {
+        using ArrayType = decltype( array );
+        auto const view = dataRepository::Wrapper< ArrayType >::cast( wrapper ).reference().toViewConst();
+        view.move( LvArray::MemorySpace::host, false );
+        for( int dim = 0; dim < ArrayType::NDIM; ++dim )
+        {
+          varStride[varIndex][dim] = view.strides()[dim];
+        }
+        inputPtrs[varIndex] = view.data();
+      } );
     }
+    totalVarSize += varSize[varIndex];
   }
 
   // Make sure the inputs do not exceed the maximum length
-  GEOSX_ERROR_IF( totalVarSize > 4, "Function input size is: " << totalVarSize );
+  GEOSX_ERROR_IF_GT_MSG( totalVarSize, MAX_VARS, "Function input size exceeded" );
 
   // Make sure the result / set size match
-  GEOSX_ERROR_IF( result.size() != set.size(), "To apply a function to a set, the size of the result and set must match" );
+  GEOSX_ERROR_IF_NE_MSG( result.size(), set.size(), "To apply a function to a set, the size of the result and set must match" );
 
-
-  forAll< serialPolicy >( set.size(), [&, set]( localIndex const i )
+  forAll< POLICY >( set.size(), [=]( localIndex const i )
   {
-    localIndex const index = set[ i ];
-    double input[4];
-    int c = 0;
-    for( int a=0; a<numVars; ++a )
+    localIndex const index = set[i];
+    real64 input[MAX_VARS]{};
+    int offset = 0;
+    for( integer varIndex = 0; varIndex < numVars; ++varIndex )
     {
-      for( int b=0; b<varSize[a]; ++b )
+      for( localIndex compIndex = 0; compIndex < varSize[varIndex]; ++compIndex )
       {
-        input[c] = input_ptrs[a][(index*varSize[a]+b)*timeVar[a]];
-        ++c;
+        input[offset++] = inputPtrs[varIndex][index * varStride[varIndex][0] + compIndex * varStride[varIndex][1]];
```

Comment:
I kept the overall structure of this function. My initial idea of moving the kernel launch inside the array type dispatch (in order to read data through the view) didn't quite work because we could have variables coming from multiple arrays. So we're still extracting and using a raw pointer for access, but now also the correct strides. This is obviously limited to 2D, supporting arbitrary dimensions this way is hard.